### PR TITLE
Jonmv/major upgrade orchestration

### PR DESCRIPTION
**Pull request by jonmv**
@mpolden please review.

### CODE REVIEW

**Comment by jonmv** [comment]
PR build timed out, unrelated.

File: component/src/test/java/com/yahoo/component/VersionIncompatibilityTest.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        assertFalse(compatibility.accept(new Version(0, 0, 0), new Version(6, 0, 0)));
+    }
+
+    @Test
+    public void testIllegalIncompatibilities() {
+        try {
+            VersionCompatibility.fromVersionList(List.of("1", "*"));
+            fail();
+        }
+        catch (IllegalArgumentException expected) { }
```
Comment (mpolden) [inline component/src/test/java/com/yahoo/component/VersionIncompatibilityTest.java:62]: Consider extracting a method for this pattern.

File: component/src/main/java/com/yahoo/component/VersionCompatibility.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * This class is instantiated with a list of versions which are only compatible with other versions
+ * that are not older than themselves. Each entry in this list is represented by a string, and any
+ * major, minor or micro with the value @{code *} means all versions are such boundaries.
+ * Versions A, B are incompatible iff. there exists an incompatibility boundary X such that
+ * (A &ge; X) &ne; (B &ge; X).
+ *
+ * @author jonmv
+ */
+public class VersionCompatibility {
```
Comment (mpolden) [inline component/src/main/java/com/yahoo/component/VersionCompatibility.java:19]: Why do we need this flexibility in verifying incompatible versions? Seeing as we use semantic versioning I don't think we ever need to express incompatibility between 1.2.3 and 1.x.x? Major is sufficient.

File: component/src/test/java/com/yahoo/component/VersionIncompatibilityTest.java (hunk 1 above)
Comment (jonmv) [reply component/src/test/java/com/yahoo/component/VersionIncompatibilityTest.java:62]: Hmm yes.

File: component/src/main/java/com/yahoo/component/VersionCompatibility.java (hunk 2 above)
Comment (jonmv) [reply component/src/main/java/com/yahoo/component/VersionCompatibility.java:19]: The intention was that we could set a different range (`List.of("*.*.*")`) for a test app, in our CD tests.

Comment (jonmv) [reply component/src/main/java/com/yahoo/component/VersionCompatibility.java:19]: If we set that for a particular application, the system will treat all its packages as incompatible with other platform versions than the one used to compile the package, and we'll have a test for this that runs all the time. I don't think there's a cleaner way to do that.

Comment (mpolden) [reply component/src/main/java/com/yahoo/component/VersionCompatibility.java:19]: I see, good idea 🙂

**Review by mpolden — APPROVED** [review]

File: component/src/main/java/com/yahoo/component/VersionCompatibility.java (hunk 2 above)
Comment (jonmv) [reply component/src/main/java/com/yahoo/component/VersionCompatibility.java:19]: Well, if there's only that special value we need, we could of course simplify this a lot ... >_<

**Pull request merged**